src/main/java/net/clgd/ccemux/rendering/tror/InputProvider.java

```diff
@@ -0,0 +1,80 @@
+package net.clgd.ccemux.rendering.tror;
+
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Scanner;
+import java.util.concurrent.BlockingDeque;
+import java.util.concurrent.LinkedBlockingDeque;
+
+import net.clgd.ccemux.emulation.EmulatedComputer;
+
+public class InputProvider {
+	private static InputProvider stdinProvider;
+	private static final Object lock = new Object();
+
+	private final Map<Integer, BlockingDeque<InputPacket>> events = new HashMap<>();
+	private final Thread thread;
+
+	public InputProvider(InputStream stream) {
+		thread = new Thread(() -> {
+			Scanner scanner = new Scanner(stream, "UTF-8");
+			while (scanner.hasNextLine()) {
+				String line = scanner.nextLine();
+				int metaStart = line.indexOf(':');
+				int metaEnd = line.indexOf(';');
+
+				if (metaStart < 0 || metaEnd < 0 || metaEnd < metaStart) continue;
+
+				String code = line.substring(0, metaStart);
+				String meta = line.substring(metaStart + 1, metaEnd);
+				String payload = line.substring(metaEnd + 1);
+
+				int computer;
+				try {
+					computer = Integer.parseInt(meta);
+				} catch (NumberFormatException e) {
+					e.printStackTrace();
+					continue;
+				}
+
+				getQueue(computer).add(new InputPacket(code, payload));
+			}
+		});
+		thread.setName("TRoR input provider");
+		thread.setDaemon(true);
+		thread.start();
+	}
+
+	public BlockingDeque<InputPacket> getQueue(EmulatedComputer computer) {
+		return getQueue(computer.getID());
+	}
+
+	private BlockingDeque<InputPacket> getQueue(int id) {
+		synchronized (events) {
```

Comment:
I'm generally not a fan of locking on publically accessible objects. Happy to changeit though.